### PR TITLE
Add example of getting DataLoader from within LightningModule

### PR DESCRIPTION
**Pull request by jgibson2**
## What does this PR do?

Updates docs with example of getting the DataLoader object from within the LightningModule, motivated by https://github.com/PyTorchLightning/pytorch-lightning/issues/10834

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [n/a] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

I did :)

### CODE REVIEW

File: docs/source/guides/data.rst [hunk 1]
```diff
@@ -357,3 +359,21 @@ option when using sequential data.
 
     # Set limit_val_batches as an int
     trainer = Trainer(limit_val_batches=100)
+
+
+*****************************************
+Accessing DataLoaders in LightningModules
+*****************************************
+
+In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the `Trainer` object:
```
Comment (rohitgr7) [inline docs/source/guides/data.rst:368]:
```suggestion
In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:
```

File: docs/source/guides/data.rst [hunk 2]
```diff
@@ -357,3 +359,21 @@ option when using sequential data.
...
+
+.. code-block:: python
+
+    def test_step(batch, batch_idx, dataloader_idx):
```
Comment (rohitgr7) [inline docs/source/guides/data.rst:372]:
```suggestion
    def test_step(self, batch, batch_idx, dataloader_idx):
```

File: docs/source/guides/data.rst [hunk 3]
```diff
@@ -357,3 +359,21 @@ option when using sequential data.
 
     # Set limit_val_batches as an int
     trainer = Trainer(limit_val_batches=100)
+
+
+*****************************************
+Accessing DataLoaders in LightningModules
+*****************************************
```
Comment (rohitgr7) [inline docs/source/guides/data.rst:366]:
```suggestion
********************************************
Accessing DataLoaders within LightningModule
********************************************
```

File: docs/source/guides/data.rst [hunk 4]
```diff
@@ -357,3 +359,21 @@ option when using sequential data.
...
+
+In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the `Trainer` object:
+
+.. code-block:: python
+
+    def test_step(batch, batch_idx, dataloader_idx):
+        test_dl = self.trainer.test_dataloaders[dataloader_idx]
```
Comment (rohitgr7) [inline docs/source/guides/data.rst:373]: I think we should add example when using `CombinedLoader` and one for train dataloader. Train dataloader is always a combined dataloader
When someone use `CombinedLoader`, the dataloaders are accessible using `self.test_dataloader.loaders`

File: docs/source/guides/data.rst [hunk 5]
```diff
@@ -357,3 +359,21 @@ option when using sequential data.
...
+        test_dataset = test_dl.dataset
+        # extract metadata, etc. from the dataset:
+        ...
+
+Since pytorch-lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata
```
Comment (rohitgr7) [inline docs/source/guides/data.rst:378]:
```suggestion
Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
```

File: docs/source/guides/data.rst (hunk 4 above)
Comment (jgibson2) [reply docs/source/guides/data.rst:373]: I'm not familiar with the internals of `CombinedLoader` -- if you don't mine providing an example I'd be happy to add it here

File: docs/source/guides/data.rst [hunk 6]
```diff
@@ -357,3 +359,20 @@ option when using sequential data.
 
     # Set limit_val_batches as an int
     trainer = Trainer(limit_val_batches=100)
+
+
+********************************************
+Accessing DataLoaders within LightningModules
+********************************************
```
Comment (awaelchli) [inline docs/source/guides/data.rst:366]:
```suggestion
*********************************************
Accessing DataLoaders within LightningModules
*********************************************
```

File: docs/source/guides/data.rst [hunk 7]
```diff
@@ -357,3 +359,20 @@ option when using sequential data.
 
     # Set limit_val_batches as an int
     trainer = Trainer(limit_val_batches=100)
+
```
Comment (awaelchli) [inline docs/source/guides/data.rst:362]:
```suggestion
----------

```

File: docs/source/guides/data.rst (hunk 4 above)
Comment (awaelchli) [reply docs/source/guides/data.rst:373]: Here is an example:

```py
def training_step(self, batch, batch_idx):
        train_dl = self.trainer.train_dataloadesr.loaders[0]
```

**Review by awaelchli — APPROVED** [review]

File: docs/source/guides/data.rst (hunk 4 above)
Comment (rohitgr7) [reply docs/source/guides/data.rst:373]: same we can extend for eval dataloaders having dict return from eval_dataloader

```py
from pytorch_lightning.trainer.supporters import CombinedLoader

def val_dataloader(self):
    val_dl1 = ...
    val_dl2 = ...
    return CombinedLoader({'dl1: val_dl1, 'dl2': val_dl2})

def validation_step(self, batch, batch_idx):
    val_dl1 = self.val_dataloaders[0]['val_dl1'] 
```

Comment (tchaton) [reply docs/source/guides/data.rst:373]: It should be self.val_dataloaders.loaders['val_dl_1']

**Review by rohitgr7 — APPROVED** [review]

File: docs/source/guides/data.rst [hunk 8]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+Accessing DataLoaders within LightningModule
+********************************************
+
+In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:
+
+.. code-block:: python
+
+    def test_step(self, batch, batch_idx, dataloader_idx):
+        test_dl = self.trainer.test_dataloaders[dataloader_idx]
+        test_dataset = test_dl.dataset
```
Comment (tchaton) [inline docs/source/guides/data.rst:265]:
```suggestion
        test_dl = self.trainer.test_dataloaders[dataloader_idx]
        test_dataset = test_dl.dataset
        test_sampler = test_dl.sampler
        ...
```

File: docs/source/guides/data.rst [hunk 9]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+        # extract metadata, etc. from the dataset:
+        ...
+
+Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
```
Comment (tchaton) [inline docs/source/guides/data.rst:269]:
```suggestion
Since PyTorch Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
```

File: docs/source/guides/data.rst [hunk 10]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+
+If you are using :class:`~pytorch_lightning.trainer.supporters.CombinedLoader`, you can access the dataloaders using the following way:
+
+If using a list of dataloaders:
```
Comment (tchaton) [inline docs/source/guides/data.rst:273]:
```suggestion
Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collection of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:

If you provided a list of dataloaders:
```

File: docs/source/guides/data.rst [hunk 11]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+
+    def test_dataloader(self):
+        test_dl1 = ...
+        test_dl2 = ...
+        return CombinedLoader([test_dl1, test_dl2])
 
-.. _sequential-data:
 
+    def test_step(self, batch, batch_idx):
+        test_dl1 = self.trainer.test_dataloaders[0].loaders[0]
```
Comment (tchaton) [inline docs/source/guides/data.rst:287]:
```suggestion
        combined_loader = self.trainer.test_dataloaders[0]
        list_of_loaders = combined_loader.loaders
        test_dl1 = list_of_loaders.loaders[0]
```

File: docs/source/guides/data.rst [hunk 12]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+
+
+If using a dictionary of dataloaders:
```
Comment (tchaton) [inline docs/source/guides/data.rst:290]:
```suggestion
If you provided dictionary of dataloaders:
```

File: docs/source/guides/data.rst [hunk 13]
```diff
@@ -252,9 +252,61 @@ set is not available at the time your model was declared. Simply pass the test s
...
+
+
+    def test_dataloader(self):
+        test_dl1 = ...
+        test_dl2 = ...
+        return CombinedLoader({"dl1": test_dl1, "dl2": test_dl2})
+
+
+    def test_step(self, batch, batch_idx):
+        test_dl1 = self.trainer.test_dataloaders[0].loaders["dl1"]
```
Comment (tchaton) [inline docs/source/guides/data.rst:304]:
```suggestion
        combined_loader = self.trainer.test_dataloaders[0]
        dictionarry_of_loaders = combined_loader.loaders
        test_dl1 = dictionarry_of_loaders["dl1"]
```

**Comment by Borda** [comment]
@jgibson2 pls use the checklist in the PR :rabbit:

File: docs/source/guides/data.rst (hunk 9 above)
Comment (rohitgr7) [reply docs/source/guides/data.rst:269]: I think just Lightning tells it's PyTorch Lightning. That's what we use at other places as well.

**Comment by jgibson2** [comment]
> @jgibson2 pls use the checklist in the PR 🐰

Sorry, what does this mean?

**Comment by akihironitta** [comment]
> > @jgibson2 pls use the checklist in the PR rabbit
> 
> Sorry, what does this mean?

@jgibson2 He's just asking you to check the boxes in the description of this PR above :) but rohitgr7 took care of that already :)

**Review by akihironitta — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

File: docs/source/guides/data.rst [hunk 14]
```diff
@@ -252,9 +252,67 @@ set is not available at the time your model was declared. Simply pass the test s
...
+    def test_dataloader(self):
+        test_dl1 = ...
+        test_dl2 = ...
+        return CombinedLoader({"dl1": test_dl1, "dl2": test_dl2})
+
+
+    def test_step(self, batch, batch_idx):
+        combined_loader = self.trainer.test_dataloaders[0]
+        dictionarry_of_loaders = combined_loader.loaders
+        test_dl1 = dictionarry_of_loaders["dl1"]
```
Comment (s-rog) [inline docs/source/guides/data.rst:310]:
```suggestion
        dictionary_of_loaders = combined_loader.loaders
        test_dl1 = dictionary_of_loaders["dl1"]
```
Just a minor typo

File: docs/source/guides/data.rst [hunk 15]
```diff
@@ -252,9 +252,67 @@ set is not available at the time your model was declared. Simply pass the test s
...
+        test_dl = self.trainer.test_dataloaders[dataloader_idx]
+        test_dataset = test_dl.dataset
+        test_sampler = test_dl.sampler
+        ...
+        # extract metadata, etc. from the dataset:
+        ...
+
+Since Lightning automatically takes care of iterating over the DataLoader, this is generally unnecessary unless you need to access metadata.
+
+Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collection of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:
```
Comment (s-rog) [inline docs/source/guides/data.rst:273]:
```suggestion
Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a :class:`~pytorch_lightning.trainer.supporters.CombinedLoader` object which supports collections of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure.:
```

File: docs/source/guides/data.rst [hunk 16]
```diff
@@ -252,9 +252,67 @@ set is not available at the time your model was declared. Simply pass the test s
...
+
+.. code-block:: python
+
+    from pytorch_lightning.trainer.supporters import CombinedLoader
+
+
+    def test_dataloader(self):
+        test_dl1 = ...
+        test_dl2 = ...
+        return CombinedLoader([test_dl1, test_dl2])
```
Comment (awaelchli) [inline docs/source/guides/data.rst:285]: The combined dataloaders are already explained above. Let's cut this short and not go into detail again here. This section should just be about accessing, not creating dataloaders.

File: docs/source/guides/data.rst [hunk 17]
```diff
@@ -252,9 +252,67 @@ set is not available at the time your model was declared. Simply pass the test s
...
 --------------
 
+********************************************
+Accessing DataLoaders within LightningModule
+********************************************
+
+In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:
+
+.. code-block:: python
+
```
Comment (Borda) [inline docs/source/guides/data.rst:262]: can we keep sense of `class` here
```suggestion
    class MyModule(LightningModule):
```

File: docs/source/guides/data.rst [hunk 18]
```diff
@@ -252,9 +252,67 @@ set is not available at the time your model was declared. Simply pass the test s
 
 --------------
 
+********************************************
+Accessing DataLoaders within LightningModule
+********************************************
+
+In the case that you require access to the DataLoader or Dataset objects, DataLoaders for each step can be accessed using the ``Trainer`` object:
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/guides/data.rst:261]: and make it tested at least with syntax

**Review by Borda — COMMENTED** [review]
please wrap all to classes and use `.. testcode::` instead of `.. code-block:: python`

**Review by tchaton — APPROVED** [review]
LGTM !
